pkg/common/consumer/managed_group.go

```diff
@@ -18,50 +18,151 @@ package consumer
+	groupErrors        chan error           // An error channel that will replicate the errors from the Sarama ConsumerGroup
 	restartWaitChannel chan struct{}        // A channel that will be closed when a stopped group is restarted
 	stopped            atomic.Value         // Boolean value indicating that the managed group is stopped
 	cancelErrors       func()               // Called by the manager's CloseConsumerGroup to terminate the error forwarding
 	cancelConsume      func()               // Called by the manager during CloseConsumerGroup
 	lockedBy           atomic.Value         // The LockToken of the ConsumerGroupAsyncCommand that requested the lock
 	cancelLockTimeout  func()               // Called internally to stop the lock timeout when a lock is removed
+	groupMutex         sync.Mutex           // Used to synchronize access to the internal sarama ConsumerGroup
 }
 
 // createManagedGroup associates a Sarama ConsumerGroup and cancel function (usually from the factory)
 // inside a new managedGroup struct.  If a timeout is given (nonzero), the lockId will be reset to an
 // empty string (i.e. "unlocked") after that time has passed.
-func createManagedGroup(logger *zap.Logger, group sarama.ConsumerGroup, cancel func()) *managedGroup {
+func createManagedGroup(ctx context.Context, logger *zap.Logger, group sarama.ConsumerGroup, cancelErrors func(), cancelConsume func()) managedGroup {
 
-	managedGrp := managedGroup{
-		logger:       logger,
-		group:        group,
-		errors:       make(chan error),
-		cancelErrors: cancel,
+	managedGrp := managedGroupImpl{
+		logger:        logger,
+		group:         group,
+		groupErrors:   make(chan error),
+		cancelErrors:  cancelErrors,
+		cancelConsume: cancelConsume,
+		groupMutex:    sync.Mutex{},
 	}
 
 	// Atomic values must be initialized with their desired type before being accessed, or a nil
 	// interface will be returned that is not of that type (string or bool in these cases)
 	managedGrp.lockedBy.Store("")   // Empty token string indicates "unlocked"
 	managedGrp.stopped.Store(false) // A managed group defaults to "started" when created
 
+	// Begin listening on the group's Errors() channel and write them to the managedGroup's errors channel
+	managedGrp.transferErrors(ctx)
+
 	return &managedGrp
 }
 
+// stop stops the managed group (which means closing the internal ConsumerGroup and marking the managed group as stopped)
+func (m *managedGroupImpl) stop() error {
+	// The managedGroup's start channel must be created (that is, the managedGroup must be marked
+	// as "stopped") before closing the internal ConsumerGroup, otherwise the consume function would
+	// return control to the factory.
+	m.createRestartChannel()
+	// Close the inner sarama ConsumerGroup, which will cause our consume() function to stop
+	// and wait for the managedGroup to start again.
+	m.groupMutex.Lock()
+	group := m.group
+	m.groupMutex.Unlock()
```

Comment:
should the close below be in the same mutex block?  if not the comment above should be moved down to that block since this block just acquires the group?

Reply:
I don't think I want to expand the scope of the mutex here; I'm just gating read-access to the m.group field (vs. writing it in the start() function), not the Close() function of the interface.  However, you are definitely correct about the comment position being unclear; I will move that closer to the Close() call.